phonenumberfmt/format_phone_number_list.py

```diff
@@ -0,0 +1,161 @@
+          adding the dialing code.
+          (This helps if we have a leading zero which implies
+           the current country's dialing prefix)
+        - Attempt validation by removing first two digits and then
+          adding the dialing code.
+          (This helps if we have two leading zeroes which
+          implies the current country's dialing prefix)
+
+    Under the hood, this uses Google's libphonenumber. If this
+    method doesn't find a phone number valid, then it must really
+    not be a valid number.
+    """
+    phone_number = str(phone_number)
+
+    if not phone_number or len(phone_number) < 3:
+        return []
+
+    dialing_prefix = dialing_prefix_for_region(implied_phone_region) or ""
+
+    valid_phone_numbers = collections.OrderedDict()
+    for number in [
+        phone_number,
+        *re.split(r"/|,", phone_number),
+        *phone_number.split(" "),
+    ]:
+        try:
+            unquoted_number = unquote(number)
+        except Exception:
+            unquoted_number = number
+        valid_numbers = []
+        _append_valid_number(
+            valid_numbers,
+            unquoted_number.split("#", maxsplit=1)[0],
+            dialing_prefix,
+            fmt,
+        )
+        _append_valid_number(
+            valid_numbers,
+            unquoted_number.split(":", maxsplit=1)[0],
+            dialing_prefix,
+            fmt,
+        )
+
+        cleaned_number = "".join(re.findall("[0-9+]+", unquoted_number))
+        _append_valid_number(valid_numbers, cleaned_number, dialing_prefix, fmt)
+        _append_valid_number(
+            valid_numbers, "+%s" % cleaned_number, dialing_prefix, fmt
+        )
+        if cleaned_number.startswith("00"):
+            _append_valid_number(
+                valid_numbers,
+                "+%s" % cleaned_number.replace("00", "", 1),
+                dialing_prefix,
+                fmt,
+            )
+        if len(valid_numbers) == 0 and not _is_already_contained(
+            valid_phone_numbers, cleaned_number
+        ):
+            _append_valid_number(
+                valid_numbers,
+                "%s%s" % (dialing_prefix, cleaned_number),
+                dialing_prefix,
+                fmt,
+            )
+            if len(valid_numbers) == 0:
+                _append_valid_number(
+                    valid_numbers,
+                    "%s%s" % (dialing_prefix, cleaned_number[1:]),
+                    dialing_prefix,
+                    fmt,
+                )
+            if len(valid_numbers) == 0:
+                _append_valid_number(
+                    valid_numbers,
+                    "%s%s" % (dialing_prefix, cleaned_number[2:]),
+                    dialing_prefix,
+                    fmt,
+                )
+
+        for valid_number in valid_numbers:
+            valid_phone_numbers[valid_number] = valid_number
+
+    return list(valid_phone_numbers.keys())
```

Comment:
What do you think about using a python generator to return the phone numbers on the fly?

Reply:
There are cases where 2 elements from the resulting list after splitting will provide the same valid number, so we need to have a set to store those valid numbers. Because of this, I don't think returning a generator is a good idea as all valid number needs to be computed before returning. And creating the generator from the resulting set won't bring any nebefits